Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/function/FunctionAppNode.java

```diff
@@ -0,0 +1,185 @@
+import com.microsoft.azuretools.telemetry.AppInsightsConstants;
+import com.microsoft.tooling.msservices.components.DefaultLoader;
+import com.microsoft.tooling.msservices.serviceexplorer.AzureRefreshableNode;
+import com.microsoft.tooling.msservices.serviceexplorer.NodeActionEvent;
+import com.microsoft.tooling.msservices.serviceexplorer.NodeActionListener;
+import com.microsoft.tooling.msservices.serviceexplorer.WrappedTelemetryNodeActionListener;
+import com.microsoft.tooling.msservices.serviceexplorer.azure.AzureNodeActionPromptListener;
+import com.microsoft.tooling.msservices.serviceexplorer.azure.appservice.file.AppServiceLogFilesRootNode;
+import com.microsoft.tooling.msservices.serviceexplorer.azure.appservice.file.AppServiceUserFilesRootNode;
+import com.microsoft.tooling.msservices.serviceexplorer.azure.webapp.base.WebAppBaseNode;
+import com.microsoft.tooling.msservices.serviceexplorer.azure.webapp.base.WebAppBaseState;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.DELETE_FUNCTION_APP;
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.FUNCTION;
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.OPEN_INBROWSER_FUNCTION_APP;
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.RESTART_FUNCTION_APP;
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.SHOWPROP_FUNCTION_APP;
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.START_FUNCTION_APP;
+import static com.microsoft.azuretools.telemetry.TelemetryConstants.STOP_FUNCTION_APP;
+
+public class FunctionAppNode extends WebAppBaseNode implements FunctionAppNodeView {
+    private static final Logger LOGGER = Logger.getLogger(FunctionAppNode.class.getName());
+    private static final String DELETE_FUNCTION_PROMPT_MESSAGE = "This operation will delete the Function App: %s.\n" +
+            "Are you sure you want to continue?";
+    private static final String DELETE_FUNCTION_PROGRESS_MESSAGE = "Deleting Function App";
+    private static final String FUNCTION_LABEL = "Function";
+
+    private final FunctionAppNodePresenter<FunctionAppNode> functionAppNodePresenter;
+    private String functionAppName;
+    private String functionAppId;
+    private String region;
+    private FunctionApp functionApp;
+
+    /**
+     * Constructor.
+     */
+    public FunctionAppNode(AzureRefreshableNode parent, String subscriptionId, FunctionApp functionApp) {
+        super(functionApp.id(), functionApp.name(), FUNCTION_LABEL, parent, subscriptionId,
+                functionApp.defaultHostName(), functionApp.operatingSystem().toString(), functionApp.state());
+        this.functionApp = functionApp;
+        this.functionAppId = functionApp.id();
+        this.functionAppName = functionApp.name();
+        this.region = functionApp.regionName();
```

Comment:
this may slow the execution of this constructor when `functionApp` is actually a `FunctionAppWrapper` 
consider make `region` a getter from `functionApp` instead a field of `FunctionAppNode`.